docs/ProductionChecklistForScalarDBCluster.md

```diff
@@ -0,0 +1,73 @@
+# Production checklist for ScalarDB Cluster
+
+This checklist provides recommendations when deploying ScalarDB Cluster in a production environment.
+
+## Before you begin
+
+In this checklist, we assume that you are deploying ScalarDB Cluster on a managed Kubernetes cluster, which is recommended.
+
+## Production checklist: ScalarDB Cluster
+
+The following is a checklist of recommendations when setting up ScalarDB Cluster in a production environment.
+
+### Number of pods and Kubernetes worker nodes
+
+To ensure that the Kubernetes cluster has high availability, you should use at least three worker nodes in three availability zones and deploy at least three pods. You also should ensure that one pod is deployed on one worker node by using [Node Affinity](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes-using-node-affinity/).
+
+### Worker node specifications
+
+From the perspective of commercial licenses, resources for one pod running ScalarDB Cluster are limited to 2vCPU / 4GB memory. In addition, we recommend deploying one ScalarDB Cluster pod and one Envoy pod on one worker node. Note that you do not need to deploy an Envoy pod when using `direct-kubernetes` mode.
+
+In other words, the following components run on one worker node:
+
+* ScalarDB Cluster pod (2vCPU / 4GB)
+* Envoy proxy (0.2–0.3 vCPU / 256–328 MB)
+* Kubernetes components
+
+With this in mind, you should use a worker node that has 4vCPU / 8GB memory resources. We recommend running only the above components on the worker node for ScalarDB Cluster.
```

Comment:
Does this mean that users should separate the Kubernetes cluster for their application pods from the Kubernetes cluster for ScalarDB Cluster?

Reply:
Could it be at least 4vCPU / 8GB memory node?
Is it good enough to run three nodes with at least  4vCPU / 8GB memory each?
Or, is there a particular reason for fixing the node spec?

Reply:
@brfrn169  
> Does this mean that users should separate the Kubernetes cluster for their application pods from the Kubernetes cluster for ScalarDB Cluster?

This means `use separate worker node` (not cluster) as follows. So, users can run their own applications in the same Kubernetes cluster (but on other worker nodes).

```mermaid
graph LR
  subgraph "Kubernetes cluster"
    subgraph "Node 1"
      A-1[ScalarDB Cluster Node]
      A-2[Envoy]
    end
    subgraph "Node 2"
      B-1[ScalarDB Cluster Node]
      B-2[Envoy]
    end
        subgraph "Node 3"
      C-1[ScalarDB Cluster Node]
      C-2[Envoy]
    end
    subgraph "Node 4"
      D-1[User App]
      D-2[Uer App]
    end
    subgraph "Node 5"
      E-1[User App]
      E-2[User App]
    end
    subgraph "Node 6"
      F-1[User App]
      F-2[User App]
    end
  end
```

Reply:
@feeblefakie  
> Could it be at least 4vCPU / 8GB memory node?
> Is it good enough to run three nodes with at least 4vCPU / 8GB memory each?
> Or, is there a particular reason for fixing the node spec?

Ah, I see. There is no reason for fixing the node spec.
These specs are calculated based on the estimated usage of each component.
Also, the number of worker nodes for ScalarDB Cluster should be the same as the number of ScalarDB Cluster pods.

So, I will update this sentence as follows.

* Add `at least`.
* Add new description `You should use the same number of worker nodes as the number of pods`.

Reply:
I updated as follows in b32f6499d8ce5b09d10fbe79d397c49c35fcdeca.

* Add `at least`.
* Add new description `You should use the same number of worker nodes as the number of pods`.
* Add new description `You should use the different worker nodes from ScalarDB Cluster pods for user application`.

Reply:
Thank you for answering my question!

Let me ask additional questions.

1. I feel that recommending `use a worker node that has at least 4vCPU / 8GB memory resources` and `use the same number of worker nodes as the number of ScalarDB Cluster pods` might result in too many number of worker nodes with very large transaction workloads. Instead, I wonder if we could host multiple ScalarDB Cluster pods on a single worker node. What do you think?
2. Using separate worker nodes for ScalarDB Cluster pods and user application pods seems to make the operation more complicated, especially in case of scaling pods. Is this a common practice in the Kubernetes world?

Reply:
I also feel having only one ScalarDB pod in each node might be too restricted and is leaned towards availability a bit too much.
Can we recommend users to have several nodes so that one node crash doesn't cause too much performance loss?

Reply:
@brfrn169  
> I feel that recommending use a worker node that has at least 4vCPU / 8GB memory resources and use the same number of worker nodes as the number of ScalarDB Cluster pods might result in too many number of worker nodes with very large transaction workloads. 

I agree with your concerns. If there is a workload that has a very large number of Tx, the number of worker nodes will be too large.

> Instead, I wonder if we could host multiple ScalarDB Cluster pods on a single worker node. What do you think?

From a technical perspective, it's possible and I think there is no problem.

> Using separate worker nodes for ScalarDB Cluster pods and user application pods seems to make the operation more complicated, especially in case of scaling pods. Is this a common practice in the Kubernetes world?

Actually, I think I have not heard of many cases like this in the Kubernetes world. In this case, we can use HPA and Cluster Autoscaler to auto-scaling pods and worker nodes. However, we have to manage some node labels for each application and ScalarDB Cluster pod. It is a bit complex operation as you mentioned. And, it has a downside that sacrificing flexibility (e.g., flexible auto-scaling).

The purpose of this recommendation is to keep high availability and high performance in the production environment. However, it might be too strict even if it is a production environment.

For your reference, I have heard some use cases that use specific worker nodes in the k8s cluster as follows.

* The specific pod wants to use a worker node that has GPU in the ML workload.
* The specific pod wants to use a worker node that has SSD in I/O heavy workload.

So, our purpose and use case might be a little peculiar.

---

@feeblefakie  

> I also feel having only one ScalarDB pod in each node might be too restricted and is leaned towards availability a bit too much.
> Can we recommend users to have several nodes so that one node crash doesn't cause too much performance loss?

I think it's a good idea. It can keep minimum availability (maybe 3 worker nodes?) and remove complex things by loosening the restriction.

Also, if we choose this way, I think we might not need to separate the worker node for the application and ScalarDB Cluster. In other words, I think we might be able to mix application pods and ScalarDB Cluster pods on the same worker node (Of course it has sufficient resources).

---

I will reconsider recommendation and update this document.
Thank you for your questions and suggestions!

Reply:
@feeblefakie @brfrn169  
According to this discussion, I updated the document as follows in 5c39b5c51959dd6298afe38e0ceb071bf95e3e4d.

* Use at least 4vCPU / 8GB memory resources for one worker node.
* Use at least three worker nodes for availability.
* Consider the worker node resources based on:
  * Resources that your application need.
  * The maximum number of pods if you use some auto scaling features.

Now, there is no too strict restriction that `1 ScalarDB Cluster pod on 1 worker node`. In other words, you can deploy several ScalarDB Cluster pods on one worker node.

Note: We have a similar description (1 ScalarDB Cluster pod on one worker node) in the manual deployment guide and sample configuration. So, I will update those documents after we will decide on the final recommendations in this PR.